### DISCUSSION TITLE
Pico-PIO-USB for post 0.15.0 TinyUSB code

### DISCUSSION THREAD

**Discussion by rppicomidi** [hathach/tinyusb#2228] [Q&A] [answered]
Although it was present in the 0.15.0 release, at the time I write this, the current master branch TinyUSB code dropped the support for the Pico-PIO-USB submodule in `hw/mcu/raspberry_pi/Pico-PIO-USB`. The family.cmake file check_and_add_pico_pio_usb_support() function still has `set(PICO_PIO_USB_PATH "${TOP}/hw/mcu/raspberry_pi/Pico-PIO-USB")` for the default. For now, I am adding Pico-PIO-USB as a git submodule to my projects that need it and am re-defining `PICO_PIO_USB_PATH` to refer to the submodule path in my project.

@hathach, is there a plan to add the Pico-PIO-USB library to `pico-sdk` somehow? If not, is it possible to add the submodule back to TinyUSB?

**Comment by hathach** [comment, marked as answer]
all submodules are revmoved and depedencies are now manage by a python script. Just run to get it

```
python3 tools/get_deps.py rp2040
```

- https://github.com/hathach/tinyusb/blob/master/tools/get_deps.py
- https://github.com/hathach/tinyusb/blob/master/docs/reference/dependencies.rst

**Comment by rppicomidi** [comment]
Thank you @hathach. What do you think about adding the following code to the RP2040 family.cmake file? The first time you build TinyUSB it will automatically fetch the dependencies for you.

```
# make sure required dependencies exist in the source tree
if ((NOT EXISTS ${TOP}/lib/FreeRTOS-Kernel) OR (NOT EXISTS ${TOP}/lib/lwip) OR (NOT EXISTS ${TOP}/tools/uf2))
    execute_process(COMMAND python3 tools/get_deps.py rp2040
    WORKING_DIRECTORY ${TOP})
endif()
```

**Reply by hathach** [reply]
yeah sure, if we put it somewhere that makes sense. Note: we should check pio-usb instead which is specific to rp2040, other like freertos lwip is kind of common depenencies.

**Reply by kilograham** [reply]
hmm; i think this should be opt in - this will cause anyone who uses `pico_sdk` to pull in FreeRTOS!

**Reply by kilograham** [reply]
(note that in the SDK, things like FreeRTOS are usually just "used" if you specify them)... certainly in the FreeRTOS case it would be good if TinyUSB could use the same FreeRTOS as the SDK build if specified.

Also, need to make sure the python3 command works on Linux, Mac, OS X (probably needs some find_package help)

**Reply by hathach** [reply]
now you mentioned this, I agree, these dependencies are meant to use with tinyusb examples only. FreeRTOS/lwip are probably already included as submodule of pico-sdk. The exception would be pico-pio-usb. So I guess, maybe just leave it as it is or only invoke this when building with one of the tinyusb examples only (not an user app).

I agree for python we should use `find_package(Python COMPONENTS Interpreter)`, thanks for bringing it up.

**Comment by harbaum** [comment]
The version pulled by the python script seems to be outdated and won't compile (due to some uint types being used). This is fixed in later versions of PIO-USB and pulling the latest master manually does indeed work.
